src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -798,19 +805,28 @@ public static void Fill<T>(T[] array, T value, int startIndex, int count)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
             }
 
-            if (startIndex < 0 || startIndex > array.Length)
+            if ((uint)startIndex > (uint)array.Length)
             {
                 ThrowHelper.ThrowStartIndexArgumentOutOfRange_ArgumentOutOfRange_Index();
             }
 
-            if (count < 0 || startIndex > array.Length - count)
+            if ((uint)count > (uint)(array.Length - startIndex))
             {
                 ThrowHelper.ThrowCountArgumentOutOfRange_ArgumentOutOfRange_Count();
             }
 
-            for (int i = startIndex; i < startIndex + count; i++)
+            if (!typeof(T).IsValueType && array.GetType() != typeof(T[]))
+            {
+                for (int i = 0; i < array.Length; i++)
+                {
+                    array[i] = value;
+                }
+            }
+            else
             {
-                array[i] = value;
+                ref T first = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(Unsafe.As<T[]>(array)), (nint)(uint)startIndex);
```

Comment:
```suggestion
                ref T first = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(array), (nint)(uint)startIndex);
```

Call to `Unsafe.As` is unnecessary: _array_ is already typed as `T[]`.

Reply:
fixedup in f261e92acee